File: 0402-remove-k-digits/0402-remove-k-digits.js
/**
 * @param {string} num
 * @param {number} k
 * @return {string}
 */
var removeKdigits = function(num, k) {
  let st = [];
  for(let i = 0;i<num.length;i++){
    while(st.length > 0 && st[st.length - 1] > num[i] && k > 0){
        st.pop();
        k--;
    }
    st.push(num[i]);
  }
  while(k>0 && st.length > 0){
    st.pop();
    k--;
  }
  let res = st.join("");
  res = res.replace(/^0+/,"");
  return res.length === 0?"0":res;
    
};

File: 0402-remove-k-digits/0402-remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:

        n = len(num)
        if n == k:
            return "0"
        stack = []
        for numi in num:
            while stack and stack[-1] > numi and k:
                stack.pop()
                k-=1
            stack.append(numi)
        if k:
            stack = stack[:len(stack)-k]
        
        result = ("").join(stack).lstrip("0")
        return result if result else "0"
        


        